File: python/src/ttio/protocols/cv_annotatable.py
"""``CVAnnotatable`` — controlled-vocabulary annotation capability."""
from __future__ import annotations

from typing import Protocol, runtime_checkable

from ..cv_param import CVParam


@runtime_checkable
class CVAnnotatable(Protocol):
    """Capability for objects that can be tagged with controlled-vocabulary
    parameters from any ontology (PSI-MS, nmrCV, CHEBI, BFO, ...).

    This is the primary extensibility mechanism in TTI-O: the schema
    stays minimal while semantic richness lives in curated external
    ontologies.

    Methods
    -------
    add_cv_param(param)
        Attach a ``CVParam`` to this object.
    remove_cv_param(param)
        Detach a previously-attached ``CVParam``. No-op if absent.
    all_cv_params()
        Return every attached ``CVParam`` in insertion order.
    cv_params_for_accession(accession)
        Return every attached ``CVParam`` whose ``accession`` matches.
    cv_params_for_ontology_ref(ontology_ref)
        Return every attached ``CVParam`` whose ``ontology_ref`` matches.
    has_cv_param_with_accession(accession)
        Return ``True`` if at least one attached ``CVParam`` matches.

    Notes
    -----
    API status: Stable.

    Cross-language equivalents
    --------------------------
    Objective-C: ``TTIOCVAnnotatable`` ·
    Java: ``com.dtwthalion.ttio.protocols.CVAnnotatable``
    """

    def add_cv_param(self, param: CVParam) -> None: ...
    def remove_cv_param(self, param: CVParam) -> None: ...
    def all_cv_params(self) -> list[CVParam]: ...
    def cv_params_for_accession(self, accession: str) -> list[CVParam]: ...
    def cv_params_for_ontology_ref(self, ontology_ref: str) -> list[CVParam]: ...
    def has_cv_param_with_accession(self, accession: str) -> bool: ...

File: java/src/main/java/global/thalion/ttio/protocols/CVAnnotatable.java
/*
 * TTI-O Java Implementation
 * Copyright (C) 2026 DTW-Thalion
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */
package global.thalion.ttio.protocols;

import global.thalion.ttio.CVParam;

import java.util.List;

/**
 * Objects implementing {@code CVAnnotatable} can be tagged with
 * controlled-vocabulary parameters from any ontology (PSI-MS, nmrCV,
 * CHEBI, BFO, ...). This is the primary extensibility mechanism in
 * TTI-O: the schema stays minimal while semantic richness lives in
 * curated external ontologies.
 *
 * <p><b>Cross-language equivalents:</b> Objective-C
 * {@code TTIOCVAnnotatable}, Python
 * {@code ttio.protocols.CVAnnotatable}.</p>
 *
 * <p><b>API status:</b> Stable (conformers may change; the protocol
 * is frozen for v0.6).</p>
 *
 *
 */
public interface CVAnnotatable {

    /**
     * Attach a {@link CVParam} to this object.
     *
     * @param param the CV parameter to attach; must not be {@code null}
     */
    void addCvParam(CVParam param);

    /**
     * Detach a previously-attached {@link CVParam}. No-op if absent.
     *
     * @param param the CV parameter to remove
     */
    void removeCvParam(CVParam param);

    /** @return every attached {@link CVParam} in insertion order. */
    List<CVParam> allCvParams();

    /**
     * @param accession the ontology accession string (e.g. {@code MS:1000514})
     * @return every attached {@link CVParam} whose {@code accession} matches
     */
    List<CVParam> cvParamsForAccession(String accession);

    /**
     * @param ontologyRef the ontology reference prefix (e.g. {@code MS})
     * @return every attached {@link CVParam} whose {@code ontologyRef} matches
     */
    List<CVParam> cvParamsForOntologyRef(String ontologyRef);

    /**
     * @param accession the ontology accession string to test
     * @return {@code true} iff at least one attached {@link CVParam} matches
     */
    boolean hasCvParamWithAccession(String accession);
}
